epitator/infection_annotator.py

```diff
@@ -0,0 +1,314 @@
+    # are plural, we assume the noun phrase is singular and add a "1" to
+    # the count metadata. Otherwise, we assume that it must be a phrase
+    # like "some patients" and do nothing.
+    if "count" not in metadata.keys() and tokens[0].dep_ == "det":
+        token_is_not_lemma = [token.lower_ != token.lemma_ for token in tokens]
+        token_is_noun = [token.pos_ == 'NOUN' for token in tokens]
+        token_is_plural = ([l and n for l, n in zip(token_is_not_lemma, token_is_noun)])
+        if not any(token_is_plural):
+            metadata["count"] = 1
+            # metadata["attributes"].append("INFERRED_FROM_SINGULAR_NC")
+
+    # "Lax metadata generation" -- so-called because it looks for tokens which
+    # are cardinal / quantity OR nummod.  This is meant to mostly cope with
+    # things in ProMED that are formatted like "193 533", which often trip up
+    # spaCy. If it finds a single token matching these criteria, it does
+    # nothing, because these are likely to be things like years or other
+    # single-token things, and this would increase the false positive rate. It
+    # handles multiple tokens in the same manner as above.
+    if "count" not in metadata.keys() and strict_only is False:
+        try:
+            lax_quant_idx = [i for (i, t) in enumerate(tokens) if t.ent_type_ in ['CARDINAL', 'QUANTITY'] or t.dep_ == 'nummod']
+            if len(lax_quant_idx) == 1:
+                count_text = tokens[lax_quant_idx[0]].text
+                metadata["count"] = parse_count_text(count_text)
+                metadata["attributes"].append(["LAX"])
+            elif len(lax_quant_idx) > 1:
+                # warning("Using lax metadata generation.")
+                # This loop groups consecutive indices into sub-lists.
+                groups = []
+                for k, g in groupby(enumerate(lax_quant_idx), lambda ix: ix[0] - ix[1]):
+                    groups.append(list(map(itemgetter(1), list(g))))
+                if len(groups) == 1:
+                    count_text = "".join([tokens[i].text for i in groups[0]])
+                    metadata["count"] = parse_count_text(count_text)
+                    metadata["attributes"].append(["JOINED_CONSECUTIVE_TOKENS", "LAX"])
+                # We will not do this.
+                # elif len(groups) > 1:
+                #     counts = []
+                #     for group in groups:
+                #         count_text = "".join([tokens[i].text for i in group])
+                #         for t in count_text:
+                #             counts.append(parse_count_text(t))
+                #     metadata["count"] = counts
+                #     metadata["attributes"].append(["MULTIPLE_COUNT_WARNING", "JOINED_CONSECUTIVE_TOKENS", "LAX"])
+        except ValueError as e:
+            metadata = {}
+    return(metadata)
+
+
+class InfectionSpan(AnnoSpan):
+    def __init__(self, source_span, include_sources=False):
```

Comment:
Agreed, it should probably not go into the init method and should go somewhere in `.annotate()`.

I have been thinking more about the points of abstraction here, and I do think that it should be broken up or abstracted, so that the `annotate` method isn't a single long block of code, because there are meaningfully different processes that occur, which might be useful to separate out.

One thought I had was to separate out the different structures that we look for into different functions or methods. For instance, right now this looks for two main formulations:

- The first is for lemmas which correspond to infections, deaths, or hospitalizations in the noun chunk, e.g. "3 cases".
- The second is to look for person lemmas in the noun chunk, and then look for verbs in a certain position in the dependency tree to indicate that they were infected or hospitalized, e.g. "a man was sickened".

Breaking these out is useful because:

1. It means that it's easier to see what bit of code generated an item, especially if I create a `metadata[token_attributes]` metadata block.
2. It gives a clear path to expansion. Adding new capabilities, e.g. new structures for infection events, or adding the capability to look for things like "suspected" etc., should go in new methods (or functions, or wherever these should go), and added to the list of methods called by `InfectionAnnotator.annotate()`

However, I'm not certain the right… break-point, if that makes sense? This code currently exists in `InfectionSpan.init()` because I kept moving it around, and this is where it was when I finished.

- Methods of `InfectionAnnotator` that are called by `annotate`. This means they're more constrained in their application, which is fine. They would take as their argument the `AnnoDoc` which is being annotated, so that they could do things like use the entire spaCy noun chunks tier as their input — and this might be useful because this means that we can look for different parts of speech as the "seeds" for annotations, depending on the method.
- Functions defined inside `infection_annotator.py`, like some of the lower-level functions which currently generate counts.
- This was also why I at one point wanted to have a `MetaGen()` class which generated metadata for a token.

In one formulation, then, `InfectionAnnotator()` might look roughly like (pseudocode):

```python
def InfectionAnnotator():
    def init(self, doc):
        self.doc = doc

    def annotate(self):
        spans = []
        spans.append(self.noun_chunks_infection_lemmas(self.doc))
        spans.append(self.noun_chunks_with_person_lemmas_and_infection_verbs(self.doc))
        spans.append(self.other_as_yet_undefined_methods(self.doc))
        doc.tiers['infections'] = AnnoTier(spans)

    def self.noun_chunks_infection_lemmas(doc):
        definition
        return(spans)

    def self.noun_chunks_with_person_lemmas_and_infection_verbs(doc):
        definition
        return(spans)

    def self.other_as_yet_undefined_methods(doc):
        definition
        return(spans)
```

Another way would see those last three methods as functions. I don't think it's worth putting them in as their own annotator subclasses.

Reply:
That pseudocode looks good. Starting with a monolithic code block usually helps to figure out where break-points should go since you can see what all the dependencies are and where code is being repeated.